### DISCUSSION TITLE
RavenDB as a windows service does not easily report failures

### DISCUSSION THREAD

**Discussion by tehkyle** [ravendb/ravendb#13244] [Q&A]
When running a RavenDB instance as a windows service and there is a error on initialization there does not appear to be a way to inspect the issue or get any meaningful feedback as to what is wrong. 

- The service manager simply reports:
    Windows could not start the <name> service on Local Computer
    Error 0xffffffff: 0xffffffff

- The event log only states:
    The <name> service entered the stopped state.
- The Raven db log contains simply:

Time,	Thread,	Level,	Source,	Logger,	Message,	Exception
     2021-12-09T21:38:15.1364155Z, 1, Operations, Server, Raven.Server.Program,  Build 42034, Version 4.2, SemVer 4.2.8, Commit d3e8007
     PID 4696, 64 bits, 4 Cores, Arch: X64
     15.999 GBytes Physical Memory, 14.131 GBytes Available Memory, 14.157 GBytes Calculated Available Memory, 0 Bytes Scratch Dirty Memory
     Using GC in server concurrent mode retaining memory from the OS.
    librvnpal Expected Version: 42012, librvnpal Current Version: 42012

In other cases, if we try to manually run .\run.ps1 in powershell we can see the error stack, but for our current issue running directly from powershell as administrator works fine.
The specific issue we are troubleshooting seems to be related to certificate permissions in win 2012 when running the service as the default "Local Service" though we have applies permissions for toe AuthenticateUser account to the machineKeys location. Running the service as Local System works, but is not the desired permission level for the DB.

If there is a way to gather insight into what the underlying exception is on start up that would be wonderful, or if anybody has run into a similar Local Service issue on win2012 please let me know.

**Comment by ayende** [comment]
The usual way this works is that the error is logged in the event viewer, aren't you seeing the details there?

**Reply by tehkyle** [reply]
This is all that I can see from service start and in event viewer. Is there a specific place I should be looking that I'm missing?

![image](https://user-images.githubusercontent.com/1372402/145616140-cf3293e6-5313-4824-9814-5bc781ce7dc0.png)

**Reply by ayende** [reply]
The event viewer is the right place. There is usually two or three events there, once of which contains the actual error.

Is this something that you can provide us with a way to reproduce? Error reporting is something that I _care_ about, and there should be something to tell you what is going on in a clear manner.

**Reply by tehkyle** [reply]
We are using a self-signed certificate to set up a single node raven cluster which has been working on other servers other than win2012. Using the same certificate in a manual setup of the raven 4.2.118 zip replicates the issue. The settings.json is the following:
```
{
  "DataDir": "RavenData",
  "License.Eula.Accepted": true,
  "Setup.Mode": "None",
  "ServerUrl": "https://127.0.0.1:8085",
  "Security.Certificate.Path": "server.certificate.igx.pfx",
  "Security.Certificate.Password": "****"
}
```

The service is set up using the provided setup-as-service.ps1 using defaults, which installs the service as Local Service.
When we start the service it throws the above popup "Error 0xffffffff: 0xffffffff"
The only error in event log viewer I can see is in System logs and reports the following:

```
The description for Event ID 7023 from source Service Control Manager cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.

If the event originated on another computer, the display information had to be saved with the event.

The following information was included with the event: 

RavenDB
%%4294967295

The locale specific resource for the desired message is not present
```

There does not appear to be any raven specific event logs in Windows Application Logs nor Applications and Services Logs.

Finally, if we switch the installed windows service to Local System instead of Local Service, it starts up just fine. The 'Authenticated User' account has been given access to the .../RSA/MachineKeys directory.

**Reply by ayende** [reply]
What happens if you run this as the Local Service from the command line? You can do that with `psexec`

**Reply by tehkyle** [reply]
Running run.ps1 returns access denied on Start-Process, running Raven.Server.exe gives:

```
System.InvalidOperationException: Unable to start the server.
 ---> System.InvalidOperationException: Unable to load the server certificate due to invalid configuration! Admin assist
ance required.
 ---> System.InvalidOperationException: Could not load certificate file C:\raven-118\Server\server.certificate.
igx.pfx
 ---> Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: Access denied.
   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(Byte[] rawData, SafePasswordHandle password, PfxCertStoreF
lags pfxCertStoreFlags)
   at Internal.Cryptography.Pal.CertificatePal.FromBlobOrFile(Byte[] rawData, String fileName, SafePasswordHandle passwo
rd, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(Byte[] rawData, String password, X509KeyStorag
eFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(Byte[] rawData, String password, X509KeyStora
geFlags keyStorageFlags)
   at Raven.Server.ServerWide.SecretProtection.LoadCertificateFromPath(String path, String password, ServerStore serverS
tore) in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\ServerWide\SecretProtection.cs:line 670
   --- End of inner exception stack trace ---
   at Raven.Server.ServerWide.SecretProtection.LoadCertificateFromPath(String path, String password, ServerStore serverS
tore) in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\ServerWide\SecretProtection.cs:line 687
   at Raven.Server.RavenServer.LoadCertificate() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\RavenServer.cs:l
ine 1388
   --- End of inner exception stack trace ---
   at Raven.Server.RavenServer.LoadCertificate() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\RavenServer.cs:l
ine 1396
   at Raven.Server.RavenServer.LoadCertificateAtStartup() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\RavenSe
rver.cs:line 1370
   --- End of inner exception stack trace ---
   at Raven.Server.RavenServer.LoadCertificateAtStartup() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\RavenSe
rver.cs:line 1374
   at Raven.Server.RavenServer.Initialize() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\RavenServer.cs:line 1
44
   at Raven.Server.Program.Main(String[] args) in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\Program.cs:line 17
5
Error during shutdown
System.AggregateException: Failed to properly close RavenServer (Object reference not set to an instance of an object.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Raven.Server.RavenServer.<Dispose>b__117_12() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\RavenServer.c
s:line 2560
   at Raven.Server.Utils.ExceptionAggregator.Execute(Action action) in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Serv
er\Utils\ExceptionAggregator.cs:line 41
   --- End of inner exception stack trace ---
   at Raven.Server.Utils.ExceptionAggregator.ThrowIfNeeded() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\Util
s\ExceptionAggregator.cs:line 71
   at Raven.Server.RavenServer.Dispose() in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\RavenServer.cs:line 2565

   at Raven.Server.Program.Main(String[] args) in C:\Builds\RavenDB-Stable-4.2\42083\src\Raven.Server\Program.cs:line 28
1
```

I would still like to be able to see these stacks without resorting to re-running the executable in an impersonated CLI.

**Reply by ayende** [reply]
Absolutely agree, and that is strange that you aren't seeing them.
For the actual error, do you have the user profile loaded?

See also: https://github.com/dotnet/runtime/issues/19774

This is the relevant line:
https://github.com/ravendb/ravendb/blob/v5.3/src/Raven.Server/ServerWide/SecretProtection.cs#L678

Note that we are using both MacineKey and Exportable, as recommended.

**Reply by tehkyle** [reply]
I am operating on the machine via RDP as administrator and opening powershell as "NT AUTHORITY\LocalService" via psexec

**Reply by tehkyle** [reply]
`For the actual error, do you have the user profile loaded?`

Forgive me, but which profile do you mean here?

**Reply by tehkyle** [reply]
We are not connecting via powershell remoting, we are using RDP. The service fails to start when running as default LocalService, but starts fine if running as LocalSystem. We can reproduce the error in powershell if we run it as "NT AUTHORITY\LocalService" using psexec. We have set security to the MachineKeys directory to allow "Authenticated User" full access. There is no logged error from the service failure, but in powershell we get an access denied exception from the cert:

---> System.InvalidOperationException: Unable to load the server certificate due to invalid configuration! Admin assist
ance required.
 ---> System.InvalidOperationException: Could not load certificate file C:\raven-118\Server\server.certificate.
igx.pfx
 ---> Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: Access denied.
   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(Byte[] rawData, SafePasswordHandle password, PfxCertStoreF
lags pfxCertStoreFlags)

Running powershell as administrator starts up without issue.

**Reply by arekpalinski** [reply]
How did you run the initial RavenDB setup wizard? What account was used then? Any chance that initially RavenDB was setup with the usage of a local admin account and then a user running RavenDB service was changed to LocalService?

**Reply by ayende** [reply]
Also, what happens if you try to open that cert as a Local Serivce from Power Shell directly?

**Reply by tehkyle** [reply]
I am able to use Get-PfxCertificate on the server.certificate pfx file as local service which matches the behavior when running powershell as admin as well:

![image](https://user-images.githubusercontent.com/1372402/146840372-f44ea051-3290-4fec-a58b-689a92014dfe.png)

Though, running .\run.ps1 as local service returns access denied on the same pfx, while admin powershell starts normally:

![image](https://user-images.githubusercontent.com/1372402/146840558-6859966c-9a26-4d99-aae4-543eed6b6c32.png)

**Reply by arekpalinski** [reply]
@tehkyle Did you see my questions in the above comment https://github.com/ravendb/ravendb/discussions/13244#discussioncomment-1832384?

**Reply by tehkyle** [reply]
My apology, I missed them. I was using an administrator command line to run "rvn.exe windows-service register" with default options, which set the service up initially as Local Service. The self-signed cert that was created for this was also done by an administrator. Additionally that cert was being rejected in the setup wizard as "invalid password" when the correct password was entered so I tried just manually configuring the settings.json.

**Reply by arekpalinski** [reply]
Okay. So calling `Get-PfxCertificate` as Local Service from Powershell works. Can you try to do load the cert from Powershell this way as Local Service:

```
> $KeyStorageFlags = [System.Security.Cryptography.X509Certificates.X509KeyStorageFlags]::Exportable -bor [System.Security.Cryptography.X509Certificates.X509KeyStorageFlags]::MachineKeySet
> $cert=New-Object System.Security.Cryptography.X509Certificates.X509Certificate2("C:\path\to\cert.pfx", "password", $KeyStorageFlags)
```

That will be literally the same call as we do here:
https://github.com/ravendb/ravendb/blob/v5.3/src/Raven.Server/ServerWide/SecretProtection.cs#L678

**Reply by tehkyle** [reply]
New-Object : Exception calling ".ctor" with "3" argument(s): "Access denied.
--

![image](https://user-images.githubusercontent.com/1372402/147125595-10ac6ab0-4dee-4ec1-a40e-d4db63bc5413.png)

**Reply by tehkyle** [reply]
as administrator:

![image](https://user-images.githubusercontent.com/1372402/147125739-1b687160-0fc7-44b4-b99a-6d79f4d88cae.png)

**Reply by arekpalinski** [reply]
I see you use Win Server 2012 and you had problems with the cert password. Maybe this is relevant for you:
https://stackoverflow.com/a/64911613

> On older windows versions, this even if the password is correct, will fail with 'access denied'.

**Reply by tehkyle** [reply]
The self signed cert we are using was created on the same machine using CreateSelfSigned on a System.Security.Cryptography.X509Certificates.CertificateRequest created with RSA and HashAlgorithmName.SHA256.

**Reply by ayende** [reply]
Is it possible that you aren't passing a password? That is what I can see from the script you shared.
Given this is a self signed cert, can you create a _new_ one and test it, then send it to us ?

**Comment by garayx** [comment]
Hi 

I have opened an issue to report the exception in the event viewer when running as a service
https://issues.hibernatingrhinos.com/issue/RavenDB-17710
